pyface/data_view/data_models/array_data_model.py

```diff
@@ -0,0 +1,261 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+""" Provides an N-dimensional array data model implementation.
+
+This module provides an a concrete implementation of a data model for an
+n-dim numpy array.
+"""
+from traits.api import Array, Instance, observe
+
+from pyface.data_view.abstract_data_model import AbstractDataModel
+from pyface.data_view.abstract_value_type import AbstractValueType
+from pyface.data_view.value_types.api import (
+    ConstantValue, FloatValue, IntValue, TextValue, no_value
+)
+from pyface.data_view.index_manager import TupleIndexManager
+
+
+class ArrayDataModel(AbstractDataModel):
+
+    #: The array being displayed.
+    data = Array()
+
+    #: The index manager that helps convert toolkit indices to data view
+    #: indices.
+    index_manager = Instance(TupleIndexManager, args=())
+
+    #: The value type of the column titles.
+    header_label_type = Instance(
+        AbstractValueType,
+        factory=ConstantValue,
+        kw={'text': "Index"},
+    )
+
+    #: The value type of the column titles.
+    column_header_type = Instance(
+        AbstractValueType,
+        factory=IntValue,
+        kw={'is_editable': False},
+    )
+
+    #: The value type of the column titles.
+    row_header_type = Instance(
+        AbstractValueType,
+        factory=IntValue,
+        kw={'is_editable': False},
+    )
+
+    #: The type of value being displayed in the data model.
+    value_type = Instance(AbstractValueType)
+
+    # Data structure methods
+
+    def get_column_count(self, row):
+        """ How many columns in the row of the data view model.
+
+        The total number of columns in the table is given by the column
+        count of the Root row.
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+
+        Returns
+        -------
+        column_count : non-negative int
+            The number of columns that the row provides.
+        """
+        return self.data.shape[-1]
```

Comment:
Because numpy arrays are regular, all rows have the same number of columns for an array model.  But we could imagine a ragged data structure where the number of columns in a particular row is different.

The root row has to return the maximum number of columns in the entire table; similarly non-leaf nodes should return the maximum number of columns of their child nodes.  I'll see if I can clarify the ABC's docstring.

Reply:
I think I share the same "I am not sure I understand what 'row' means" sentiment...

> The root row has to return the maximum number of columns in the entire table; similarly non-leaf nodes should return the maximum number of columns of their child nodes.

This sentence is consistent with my current understanding of what a row is... I think I have so far passed the *is* tests, until I read the implementation of this method here, so maybe I misunderstood it afterall.

Suppose the array is a 3-D array, of shape `(10, 4, 5)`, why should `get_column_count` always return 5?  This could either point to my misunderstanding what "row" means, or this is a bug in this implementation?

Reply:
Nope not a bug, but I guess at this point I am confused about the confusion.  What other value do you think it should return?  That might help me understand how to clarify.

Reply:
Once you confirmed this is not a bug I think I can fit it in my mental model again. But I think we can agree that the description does not allow that understanding to occur at the beginning.

Let me try to walk-through an example and perhaps you can help me to verify my understanding.
Suppose we have an array `data` of shape `(10, 9, 8, 7, 6, 5)`.
The "root row" should see 10 rows.
Suppose the `row` is `(1,)` (I am using tuple because it fits my model better than using list, see comment above), then the data for this row is `data[1]`, which has a shape of `(9, 8, 7, 6, 5)`.  The nested table should have 9 rows. The concept of columns isn't very well-defined at this point because the array is still not a 1D.
Say we then expand this row, and we get `row = (1, 2)`, the data is `data[1, 2]`, which has a shape of `(8, 7, 6, 5)`, the nested table for this `row` will have 8 rows, again the number of columns is still not very well-defined...
We carry on with expanding the "rows", until we get to, say `row = (1, 2, 3, 4, 5)`, corresponding to `data[1, 2, 3, 4, 5]`, which has a shape of `(5,)`. Here is where the number of columns is well-defined, it is 5.

I guess the issue here is naming... for intermediate "row" that are not the "leaf row", they aren't really rows visually, they are nested "tables". In these nested tables, or "non-leaf nodes", "number of columns" isn't very well defined, let alone "maximum number of columns".

I don't have a suggestion for different names yet... I wanted to confirm my understanding first.

Reply:
Perhaps some of the confusion is because `(1,)` _is_ an actual displayed row in the view, separate from (for example) `(1, 0)`.  In other words it _is_ a visual row, and it can have data associated with it - for example the cells in the row could show aggregates (sums, avergaes, etc.) of the values in the appropriate column.  We just haven't done this for any of the examples.

The only "row" which isn't a displayed row is the root `()`, but we use that to access the column headers which works well, since in some sense the column headers are the cells that are at the root of each column.

Reply:
Perhaps this illustration might make the indexing scheme clearer:

<img width="720" alt="Screen Shot 2020-07-01 at 10 25 11 am" src="https://user-images.githubusercontent.com/600761/86229028-0bbaec00-bb87-11ea-8443-fa7830f43ce5.png">

Reply:
Ah... I think I was missing this :)

Just to check: Could I implement an `AbstractDataModel` for a N-D array, of shape `(10, 9, 8, 7, 6, 5)` (again, say) like this:

```
    data = Array()

    def get_column_count(self, row):
         return {
               0: 0,  # root row is going to show a nothing, don't want to bother doing calculation
               1: 2,  # the first slice, e.g. `data[1]` is going to show the mean and the variance of all the values
               2: 0,  # we will show an empty row again for this slice, why not?
               ...
         }[len(row)]

    def get_value(self, row, column):
        if len(row) == 0:
            return "Title"
        if len(column) == 0:
            return "Group #{}".format(len(row))

        column_index, = column
        if len(row) == 1:
            if column_index == 0:
                 return self.data[row].mean()
            return self.data[row].var()
        if len(row) == 2:
            return # will we get here if the number of columns is zero?
        ...
```

Reply:
I've dug into this a bit, and it looks like although the Qt `QAbstractItemModel` _claims_ to allow this sort of thing in principle:
> The model has a rowCount() and a columnCount() for each level of the hierarchy.
In practice (ie. in all actual views) it only takes into consideration the number of rows returned when called with the root.  ie. in the above example you would get only the index column, because the root row returns returns 0; if you instead had the root row returning 100, then _all_ rows get queried for all 100 columns, even if they claim to support less.  Which is bad, if people think that they only need to return 2 values.  This may be a bug in Qt, or I may have made a mistake somewhere, but I can't see an issue with the experiments I was trying.  I'll play around with this a bit more to see if there is something I'm missing in Qt model.

So I'm leaning towards changing `get_column_count` to not take an argument, and just return the number of columns in the tabular view, and if a row wants show less, then they can return null values and appropriate null value types for the cells they don't want to populate.

Reply:
I've gone and made column counts constant.  If we find an example where we really want ragged columns, we can revisit.